# Buổi 02 – Phân tích dữ liệu bán hàng (Sales Data Sample)

**Bản giải đề xuất** – code đơn giản, mỗi câu chỉ 1 vài dòng lệnh trực tiếp.

> **Lưu ý:** File đề gốc không có "Phần 5" và Câu 18–21 (nhảy thẳng Câu 17 sang Câu 22 ở "Phần 6"). Nên hỏi lại giảng viên nếu cần bổ sung.

---
## Phần 1: Nhập và khám phá dữ liệu tổng quan

In [2]:
import pandas as pd

In [3]:
df = pd.read_excel('sales_data_sample.xlsx')

**Câu 1.** Đọc dữ liệu, hiển thị 5 dòng đầu và 5 dòng cuối.

In [4]:
display(df.head(5))
df.tail(5)

,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit
0,10107,30,2,2871.00,2003-02-24,Shipped,USA,Small,0.00,41.9136
1,10121,34,5,2765.90,2003-05-07,Shipped,France,Small,0.00,219.5820
2,10134,41,2,3884.34,2003-07-01,Shipped,France,Medium,0.00,6.8714
3,10145,45,6,3746.70,2003-08-25,Shipped,USA,Medium,0.45,-383.0310
4,10159,49,14,5205.27,2003-10-10,Shipped,USA,Medium,0.20,2.5164


,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit
2818,10350,20,15,2244.40,2004-12-02,Shipped,Spain,Small,0.0,315.7404
2819,10373,29,1,3978.51,2005-01-31,Shipped,Finland,Medium,0.2,2.3520
2820,10386,43,4,5417.57,2005-03-01,Resolved,Spain,Medium,0.2,7.4694
2821,10397,34,1,2116.16,2005-03-28,Shipped,France,Small,0.2,22.6782
2822,10414,47,9,3079.44,2005-05-06,On Hold,USA,Medium,0.2,0.5580


**Câu 2.** Số dòng/cột và dtype từng cột.

In [5]:
df.shape

(2823, 10)

In [6]:
df.dtypes

ORDERNUMBER                 int64
QUANTITYORDERED             int64
ORDERLINENUMBER             int64
SALES                     float64
ORDERDATE          datetime64[us]
STATUS                        str
COUNTRY                       str
DEALSIZE                      str
Discount                  float64
Profit                    float64
dtype: object

**Câu 3.** Kiểm tra missing values và duplicate.

In [7]:
df.isnull().sum()

ORDERNUMBER        0
QUANTITYORDERED    0
ORDERLINENUMBER    0
SALES              0
ORDERDATE          0
STATUS             0
COUNTRY            0
DEALSIZE           0
Discount           0
Profit             0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

**Câu 4.** `describe()` cho các cột số.

In [9]:
df[['QUANTITYORDERED', 'SALES', 'Discount', 'Profit']].describe()

,QUANTITYORDERED,SALES,Discount,Profit
count,2823.000000,2823.000000,2823.000000,2823.000000
mean,35.092809,3553.889072,0.152710,25.383894
std,9.741443,1841.865106,0.203333,192.452771
min,6.000000,482.130000,0.000000,-3839.990400
25%,27.000000,2203.430000,0.000000,1.801600
50%,35.000000,3184.800000,0.150000,8.880000
75%,43.000000,4508.000000,0.200000,30.488300
max,97.000000,14082.800000,0.800000,3919.988800


**Nhận xét:** `Discount` từ 0–0.8, trung bình ~0.15 (15%). `Profit` dao động rất rộng, từ -3839.99 đến 3919.99, trung bình chỉ ~25.38 — lợi nhuận giữa các đơn rất không đồng đều.

**Câu 5.** Số quốc gia, STATUS, DEALSIZE khác nhau và tên các giá trị.

In [10]:
df['COUNTRY'].nunique(), df['COUNTRY'].unique()

(19,
 <StringArray>
 [        'USA',      'France',      'Norway',   'Australia',     'Finland',
      'Austria',          'UK',       'Spain',      'Sweden',   'Singapore',
       'Canada',       'Japan',       'Italy',     'Denmark',     'Belgium',
  'Philippines',     'Germany', 'Switzerland',     'Ireland']
 Length: 19, dtype: str)

In [11]:
df['STATUS'].nunique(), df['STATUS'].unique()

(6,
 <StringArray>
 ['Shipped', 'Disputed', 'In Process', 'Cancelled', 'On Hold', 'Resolved']
 Length: 6, dtype: str)

In [12]:
df['DEALSIZE'].nunique(), df['DEALSIZE'].unique()

(3,
 <StringArray>
 ['Small', 'Medium', 'Large']
 Length: 3, dtype: str)

---
## Phần 2: Lọc và sắp xếp dữ liệu

**Câu 6.** Lọc SALES > 5000, đếm số đơn.

In [13]:
df[df['SALES'] > 5000]

,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit
4,10159,49,14,5205.27,2003-10-10,Shipped,USA,Medium,0.2,2.5164
7,10188,48,1,5512.32,2003-11-18,Shipped,Norway,Medium,0.2,90.7152
20,10341,41,9,7737.93,2004-11-24,Shipped,Austria,Large,0.2,7.3840
25,10417,66,2,7516.08,2005-05-13,Disputed,Spain,Large,0.2,4.2224
26,10103,26,11,5404.62,2003-01-29,Shipped,Norway,Medium,0.0,11.7741
...,...,...,...,...,...,...,...,...,...,...
2685,10361,44,10,5001.92,2004-12-17,Shipped,Australia,Medium,0.2,29.9850
2686,10375,44,11,5208.72,2005-02-03,Shipped,France,Medium,0.7,-2.4564
2689,10401,77,9,7084.00,2005-04-03,On Hold,USA,Large,0.2,1.6400
2715,10397,48,3,5192.64,2005-03-28,Shipped,France,Medium,0.0,62.3904


In [14]:
len(df[df['SALES'] > 5000])

549

**Câu 7.** Lọc USA + DEALSIZE Large, sắp xếp SALES giảm dần, lấy 10 dòng đầu.

In [15]:
df[(df['COUNTRY'] == 'USA') & (df['DEALSIZE'] == 'Large')].sort_values('SALES', ascending=False).head(10)

,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit
598,10407,76,2,14082.80,2005-04-22,On Hold,USA,Large,0.2,1.0178
744,10322,50,6,12536.50,2004-11-04,Shipped,USA,Large,0.2,1.6762
44,10312,48,3,11623.70,2004-10-21,Shipped,USA,Large,0.0,19.7714
1133,10333,46,2,11336.70,2004-11-18,Shipped,USA,Large,0.2,-23.3640
188,10127,46,2,11279.20,2003-06-03,Shipped,USA,Large,0.2,6.2550
2505,10388,46,2,10066.60,2005-03-03,Shipped,USA,Large,0.0,327.5060
131,10400,64,9,9661.44,2005-04-01,Shipped,USA,Large,0.2,29.3640
175,10276,50,3,9631.00,2004-08-02,Shipped,USA,Large,0.2,6.5520
319,10142,46,11,9470.94,2003-08-08,Shipped,USA,Large,0.0,80.0226
164,10147,48,7,9245.76,2003-09-05,Shipped,USA,Large,0.2,2.7324


**Câu 8.** Lọc Profit âm, value_counts STATUS và DEALSIZE.

In [16]:
df_profit = df[df['Profit'] < 0]
print(df_profit['STATUS'].value_counts())
print(df_profit['DEALSIZE'].value_counts())

STATUS
Shipped       454
In Process     14
Cancelled      14
On Hold        10
Resolved        7
Disputed        3
Name: count, dtype: int64
DEALSIZE
Medium    241
Small     228
Large      33
Name: count, dtype: int64


**Nhận xét:** Lỗ tập trung nhiều nhất ở `STATUS = Shipped` (454 đơn) và `DEALSIZE = Medium/Small`, không phải do huỷ/tranh chấp.

**Câu 9.** Lọc Discount >= 0.3, so sánh SALES trung bình với toàn bộ.

In [36]:
df[df['Discount'] >= 0.3]['SALES'].mean()

np.float64(3590.8555091383814)

---
## Phần 3: Tạo cột mới và biến đổi dữ liệu

**Câu 10.** Tạo `ORDER_YEAR`, `ORDER_MONTH`, `ORDER_QUARTER` từ `ORDERDATE`.

In [19]:
df['ORDER_YEAR'] = df['ORDERDATE'].dt.year
df['ORDER_MONTH'] = df['ORDERDATE'].dt.month
df['ORDER_QUARTER'] = df['ORDERDATE'].dt.quarter
df.head()

,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit,ORDER_YEAR,ORDER_MONTH,ORDER_QUARTER
0,10107,30,2,2871.00,2003-02-24,Shipped,USA,Small,0.00,41.9136,2003,2,1
1,10121,34,5,2765.90,2003-05-07,Shipped,France,Small,0.00,219.5820,2003,5,2
2,10134,41,2,3884.34,2003-07-01,Shipped,France,Medium,0.00,6.8714,2003,7,3
3,10145,45,6,3746.70,2003-08-25,Shipped,USA,Medium,0.45,-383.0310,2003,8,3
4,10159,49,14,5205.27,2003-10-10,Shipped,USA,Medium,0.20,2.5164,2003,10,4


**Câu 11.** Tạo `PROFIT_MARGIN`, tìm đơn cao nhất/thấp nhất.

In [40]:
df['PROFIT_MARGIN'] = df['Profit'] / df['SALES']

In [41]:
df.sort_values('PROFIT_MARGIN', ascending=False).head(1)

,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit,ORDER_YEAR,ORDER_MONTH,ORDER_QUARTER,PROFIT_MARGIN,NET_SALES
509,10193,46,6,2455.02,2003-11-21,Shipped,Australia,Small,0.0,3177.475,2003,11,4,1.294277,2455.02


In [22]:
df.sort_values('PROFIT_MARGIN').head(1)

,ORDERNUMBER,QUANTITYORDERED,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,COUNTRY,DEALSIZE,Discount,Profit,ORDER_YEAR,ORDER_MONTH,ORDER_QUARTER,PROFIT_MARGIN
683,10149,24,10,1496.64,2003-09-12,Shipped,USA,Small,0.5,-3839.9904,2003,9,3,-2.565741


**Câu 12.** Tạo `NET_SALES`, so sánh tổng SALES và tổng NET_SALES.

In [23]:
df['NET_SALES'] = df['SALES'] * (1 - df['Discount'])

In [24]:
df['SALES'].sum(), df['NET_SALES'].sum()

(np.float64(10032628.85), np.float64(8470955.407699998))

---
## Phần 4: Nhóm dữ liệu (Groupby) và tổng hợp

**Câu 13.** Tổng SALES và Profit theo COUNTRY, Top 5 doanh số.

In [25]:
df.groupby(by='COUNTRY')['SALES'].sum()

COUNTRY
Australia       630623.10
Austria         202062.53
Belgium         108412.62
Canada          224078.56
Denmark         245637.15
Finland         329581.91
France         1110916.52
Germany         220472.09
Ireland          57756.43
Italy           374674.31
Japan           188167.81
Norway          307463.70
Philippines      94015.73
Singapore       288488.41
Spain          1215686.92
Sweden          210014.21
Switzerland     117713.56
UK              478880.46
USA            3627982.83
Name: SALES, dtype: float64

In [26]:
df.groupby(by='COUNTRY')[['SALES', 'Profit']].sum()

,SALES,Profit
COUNTRY,,
Australia,630623.10,8113.6187
Austria,202062.53,655.5536
Belgium,108412.62,1143.8593
Canada,224078.56,5822.9162
Denmark,245637.15,2008.7175
Finland,329581.91,-378.7674
France,1110916.52,2940.2749
Germany,220472.09,1677.1145
Ireland,57756.43,486.0154


In [27]:
df.groupby('COUNTRY')['SALES'].sum().sort_values(ascending=False).head(5)

COUNTRY
USA          3627982.83
Spain        1215686.92
France       1110916.52
Australia     630623.10
UK            478880.46
Name: SALES, dtype: float64

**Câu 14.** SALES và Profit trung bình theo DEALSIZE.

In [28]:
df.groupby('DEALSIZE')[['SALES', 'Profit']].mean()

,SALES,Profit
DEALSIZE,,
Large,8293.753248,8.481640
Medium,4398.433699,28.543171
Small,2061.682800,24.043187


**Nhận xét:** Large có SALES trung bình cao nhất nhưng Profit trung bình lại thấp nhất; Medium mới là nhóm lãi trung bình cao nhất.

**Câu 15.** Số đơn duy nhất theo STATUS, tỷ lệ % Cancelled/Disputed.

In [29]:
df.groupby('STATUS')['ORDERNUMBER'].nunique()

STATUS
Cancelled       4
Disputed        3
In Process      6
On Hold         4
Resolved        4
Shipped       286
Name: ORDERNUMBER, dtype: int64

In [30]:
df.groupby('STATUS')['ORDERNUMBER'].nunique()[['Cancelled', 'Disputed']].sum() / df['ORDERNUMBER'].nunique() * 100

np.float64(2.2801302931596092)

**Câu 16.** `pivot_table` tổng SALES theo COUNTRY và DEALSIZE.

In [31]:
pd.pivot_table(df, values='SALES', index='COUNTRY', columns='DEALSIZE', aggfunc='sum')

DEALSIZE,Large,Medium,Small
COUNTRY,,,
Australia,60252.04,381245.00,189126.06
Austria,32792.08,125314.10,43956.35
Belgium,NaN,71531.37,36881.25
Canada,9064.89,137430.48,77583.19
Denmark,49886.67,140050.70,55699.78
Finland,40012.32,201741.96,87827.63
France,179004.87,643606.22,288305.43
Germany,30496.06,133151.79,56824.24
Ireland,15439.44,26822.87,15494.12


**Câu 17.** Tổng SALES theo ORDER_YEAR, năm nào cao nhất.

In [32]:
df.groupby('ORDER_YEAR')['SALES'].sum()

ORDER_YEAR
2003    3516979.54
2004    4724162.60
2005    1791486.71
Name: SALES, dtype: float64

In [33]:
df.groupby('ORDER_YEAR')['ORDER_MONTH'].nunique()

ORDER_YEAR
2003    12
2004    12
2005     5
Name: ORDER_MONTH, dtype: int64

**Nhận xét:** Doanh số cao nhất là năm 2004. Năm 2005 chỉ có dữ liệu đến tháng 5 (5/12 tháng) nên không so sánh trực tiếp với 2003/2004 được.

---
## Phần 6: Câu hỏi tổng hợp / thảo luận

**Câu 22.** Từ những phân tích trên, hãy viết một đoạn nhận xét ngắn (5–7 dòng) về:
- Quốc gia nào đang mang lại doanh số/lợi nhuận tốt nhất cho công ty ABC?
- Nhóm `DEALSIZE` nào hiệu quả nhất?
- Có rủi ro nào đáng lưu ý không?
- Đề xuất 1–2 hành động.

**Nhận xét tổng hợp:** USA mang lại doanh số/lợi nhuận cao nhất, gấp ~3 lần Spain (đứng thứ 2). Nhóm Medium hiệu quả về lợi nhuận trung bình nhất, dù Large có giá trị đơn hàng cao hơn. Rủi ro: nhiều đơn Shipped vẫn bị lỗ, tỷ lệ Cancelled/Disputed ~2.3%. Đề xuất: rà soát chính sách chiết khấu cho các đơn lỗ, và tập trung nguồn lực vào USA/Spain.